File: backend/src/routes/auth/bind.rs
//! 绑定匿名身份相关路由

use rocket::serde::json::Json;
use rocket::serde::{Deserialize, Serialize};
use rocket::State;
use mongodb::Database;

use crate::config::OAuthConfig;
use crate::guards::AuthGuard;
use crate::models::{ApiResponse, Reader, ReaderResponse};
use crate::services::{ReaderRepository, AccountRepository};
use crate::services::auth::identity::IdentityService;

/// 绑定匿名身份请求
#[derive(Debug, Deserialize, Serialize)]
pub struct BindAnonymousRequest {
    pub name: String,
    pub email: String,
}

#[post("/bind-anonymous", data = "<request>")]
pub async fn bind_anonymous_identity(
    auth: AuthGuard,
    request: Json<crate::routes::auth::bind::BindAnonymousRequest>,
    db: &State<Database>,
    config: &State<OAuthConfig>,
) -> Result<Json<ApiResponse<ReaderResponse>>, Json<ApiResponse<()>>> {
    let id_service = IdentityService::new(db, config.jwt_secret.clone());
    let reader_repo = ReaderRepository::new(db);

    let anon_reader = reader_repo.find_by_name_and_email(&request.name, &request.email).await
        .map_err(|_| ApiResponse::internal_error("数据库查询失败".to_string()))?
        .ok_or_else(|| ApiResponse::not_found("未匹配到匿名身份".to_string()))?;

    // 迁移所有 Account
    id_service.merge_identities(auth.user_id, anon_reader.id).await
        .map_err(|e| ApiResponse::bad_request(e))?;

    // 删除临时的 Reader（如果有）
    let _ = reader_repo.delete_reader(auth.user_id).await;

    let token = id_service.issue_token(anon_reader.id, anon_reader.is_owner)
        .map_err(|e| ApiResponse::internal_error(e))?;

    Ok(ApiResponse::json_success_with_message(anon_reader.into(), token))
}

#[post("/skip-bind")]
pub async fn skip_bind(auth: AuthGuard, db: &State<Database>, config: &State<OAuthConfig>) -> Result<Json<ApiResponse<ReaderResponse>>, Json<ApiResponse<()>>> {
    let reader_repo = ReaderRepository::new(db);
    let account_repo = AccountRepository::new(db);
    let id_service = IdentityService::new(db, config.jwt_secret.clone());

    // 如果已经有 Reader 了，直接返回
    if let Ok(Some(r)) = reader_repo.find_by_id(auth.user_id).await {
        return Ok(ApiResponse::json_success(r.into()));
    }

    // 从 Account 创建 Reader
    let accounts = account_repo.find_by_user_id(auth.user_id).await.map_err(|_| ApiResponse::internal_error("查询失败".to_string()))?;
    let acc = accounts.first().ok_or_else(|| ApiResponse::not_found("未找到账号信息".to_string()))?;

    let is_first = reader_repo.is_empty().await.unwrap_or(false);
    let new_reader = Reader {
        id: bson::oid::ObjectId::new(),
        email: acc.oauth_email.clone().unwrap_or_default(),
        name: acc.oauth_name.clone().unwrap_or_else(|| "用户".to_string()),
        handle: acc.oauth_handle.clone().unwrap_or_else(|| Reader::generate_handle("user")),
        image: acc.oauth_avatar.clone().unwrap_or_default(),
        is_owner: is_first,
        email_verified: Some(true),
        created_at: bson::DateTime::now(),
        updated_at: bson::DateTime::now(),
    };

    let new_id = reader_repo.create_reader(&new_reader).await.map_err(|_| ApiResponse::internal_error("创建失败".to_string()))?;
    id_service.merge_identities(auth.user_id, new_id).await.map_err(|e| ApiResponse::internal_error(e))?;

    let token = id_service.issue_token(new_id, is_first).map_err(|e| ApiResponse::internal_error(e))?;
    Ok(ApiResponse::json_success_with_message(new_reader.into(), token))
}

#[get("/bindable-identities")]
pub async fn get_bindable_identities(auth: AuthGuard, db: &State<Database>) -> Result<Json<ApiResponse<Vec<ReaderResponse>>>, Json<ApiResponse<()>>> {
    let account_repo = AccountRepository::new(db);
    let reader_repo = ReaderRepository::new(db);

    let accounts = account_repo.find_by_user_id(auth.user_id).await.unwrap_or_default();
    let emails: Vec<String> = accounts.iter().filter_map(|a| a.oauth_email.clone()).collect();

    let all_readers = reader_repo.get_all().await.map_err(|_| ApiResponse::internal_error("数据库失败".to_string()))?;
    let bindable = all_readers.into_iter()
        .filter(|r| r.id != auth.user_id && emails.contains(&r.email))
        .map(Into::into).collect();

    Ok(ApiResponse::json_success(bindable))
}

File: backend_ts/src/routes/auth/bind.ts
import { Elysia, t } from "elysia";
import { ObjectId } from "mongodb";
import { getConfig } from "@/config";
import { db_delete, db_find, db_read, db_update } from "@/lib/db";
import { authMiddleware, generateToken, jwtPlugin } from "@/middleware/auth";

const config = getConfig();
const DB_NAME = config.mongodb.database || "mx-space";

/**
 * Auth 绑定操作路由
 *
 * 提供以下端点：
 * - POST /auth/bind-anonymous - 绑定匿名身份
 * - POST /auth/skip-bind - 跳过绑定，创建新 Reader
 * - GET /auth/bindable-identities - 获取可绑定的身份列表
 */
export const bindRoutes = new Elysia()
  .use(authMiddleware())
  .use(jwtPlugin())

  // POST /auth/bind-anonymous - 绑定匿名身份
  .post(
    "/bind-anonymous",
    async ({ body, user, jwt, success, error, set }) => {
      try {
        if (!user) {
          set.status = 401;
          return error(401, "Authentication required");
        }

        // 查找匿名 Reader
        const anonReader = await db_find(DB_NAME, "readers", {
          name: body.name,
          mail: body.email,
        });

        if (!anonReader) {
          set.status = 404;
          return error(404, "未匹配到匿名身份");
        }

        // 迁移所有 Account（将临时用户的账号转移到匿名用户）
        await db_update(
          DB_NAME,
          "accounts",
          { userId: new ObjectId(user.userId) },
          { userId: anonReader._id },
        );

        // 删除临时的 Reader
        await db_delete(DB_NAME, "readers", { _id: new ObjectId(user.userId) });

        // 生成新 token
        const token = await generateToken(jwt, {
          userId: anonReader._id.toString(),
          email: anonReader.mail,
          role: anonReader.isOwner ? "admin" : "user",
          isOwner: anonReader.isOwner,
        });

        return success(
          {
            ...anonReader,
            token,
          },
          200,
        );
      } catch (err) {
        console.error("Error binding anonymous identity:", err);
        set.status = 500;
        return error(500, "Failed to bind anonymous identity");
      }
    },
    {
      body: t.Object({
        name: t.String(),
        email: t.String(),
      }),
    },
  )

  // POST /auth/skip-bind - 跳过绑定
  .post("/skip-bind", async ({ user, jwt, success, error, set }) => {
    try {
      if (!user) {
        set.status = 401;
        return error(401, "Authentication required");
      }

      // 检查是否已经有 Reader
      const existingReader = await db_find(DB_NAME, "readers", {
        _id: new ObjectId(user.userId),
      });

      if (existingReader) {
        return success(existingReader);
      }

      // 从 Account 获取信息创建 Reader
      const accounts = await db_read(DB_NAME, "accounts", {
        userId: new ObjectId(user.userId),
      });

      if (!accounts || accounts.length === 0) {
        set.status = 404;
        return error(404, "未找到账号信息");
      }

      const account = accounts[0];

      // 检查是否是第一个用户（站长）
      const allReaders = await db_read(DB_NAME, "readers", {});
      const isFirst = allReaders.length === 0;

      // 创建新 Reader
      const newReader = {
        _id: new ObjectId(user.userId),
        mail: account.oauthEmail || "",
        name: account.oauthName || "用户",
        handle: account.oauthHandle || `user_${Date.now()}`,
        image: account.oauthAvatar || "",
        isOwner: isFirst,
        emailVerified: true,
        createdAt: new Date(),
        updatedAt: new Date(),
      };

      // 插入 Reader（使用 update with upsert）
      await db_update(DB_NAME, "readers", { _id: newReader._id }, newReader);

      // 生成新 token
      const token = await generateToken(jwt, {
        userId: newReader._id.toString(),
        email: newReader.mail,
        role: isFirst ? "admin" : "user",
        isOwner: isFirst,
      });

      return success({
        ...newReader,
        token,
      });
    } catch (err) {
      console.error("Error skipping bind:", err);
      set.status = 500;
      return error(500, "Failed to skip bind");
    }
  })

  // GET /auth/bindable-identities - 获取可绑定的身份列表
  .get("/bindable-identities", async ({ user, success, error, set }) => {
    try {
      if (!user) {
        set.status = 401;
        return error(401, "Authentication required");
      }

      // 获取当前用户的所有账号
      const accounts = await db_read(DB_NAME, "accounts", {
        userId: new ObjectId(user.userId),
      });

      // 提取所有邮箱
      const emails = accounts
        .map((a: any) => a.oauthEmail)
        .filter((email: string) => email);

      if (emails.length === 0) {
        return success([]);
      }

      // 查找所有匹配邮箱的 Reader（排除当前用户）
      const allReaders = await db_read(DB_NAME, "readers", {});
      const bindableReaders = allReaders.filter(
        (r: any) =>
          r._id.toString() !== user.userId && emails.includes(r.mail),
      );

      return success(bindableReaders);
    } catch (err) {
      console.error("Error fetching bindable identities:", err);
      set.status = 500;
      return error(500, "Failed to fetch bindable identities");
    }
  });
